File: java/src/atcoder/ABC099/C.java
package atcoder.ABC099;

import java.util.*;
import java.util.stream.Collectors;

public class C {

    /*
    ある銀行では、お金の引き出しを難しくするために、一回の操作で引き出せる金額が以下のいずれかとなっています。
    1 円
    6 円、6^2(=36) 円、6^3(=216) 円、…
    9 円、9^2(=81) 円、9^3(=729) 円、…
    この銀行からちょうど N 円を引き出すには少なくとも何回の操作が必要か求めてください。
    ただし、一度引き出したお金を再び預け入れてはならないとします。
     */

    static int N;

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        N = sc.nextInt();

//        solve();

        test();
    }


    public static void solve() {

        // 6の累乗を持っておく
//        int[] pow6 = new int[10];
//        int[] pow9 = new int[10];
//        for (int i = 0; i < 100; i++) {
//            int p6 = (int)Math.pow(6, i+1);
//            if (p6 > 100000) {
//                break;
//            }
//            pow6[i] = p6;
//
//            int p9 = (int)Math.pow(9, i+1);
//            if (p9 < 100000) {
//                pow9[i] = p9;
//            }
//        }

        List<Integer> list = new ArrayList<>();
        for (int i = 0; i < 100; i++) {
            int p6 = (int)Math.pow(6, i+1);
            if (p6 > 100000) {
                break;
            }
            list.add(p6);

            int p9 = (int)Math.pow(9, i+1);
            if (p9 < 100000) {
                list.add(p9);
            }
        }
        Collections.sort(list);



        System.out.println(list);
    }

    public static void test() {
        int min = N;
        for (int i = 0; i <= N; i++) {
            int count = 0;
            int t = i;
            while (t > 0) {
                count += t % 6;
                t /= 6;
            }

            t = N - i;

            while (t > 0) {
                count += t % 9;
                t /= 9;
            }

            min = Math.min(min, count);
        }
        System.out.println(min);
    }
}

File: kotlin/src/atcoder/ABC099/C.kt
package atcoder.ABC099

fun main(args: Array<String>) {
    var N = readLine()!!.toInt()


    // 「大きいほうから引いていく」の貪欲法じゃだめ
    // （例）12円のとき
    //  最大の9円を引いてしまうと、残りは1円を3回引かないといけないので結果4回かかる
    //  6円6円とひけば2回ですむ

    // N円のときの求める最小回数をdp[N]として考える
    //    dp[0] = 0
    //    dp[N] = min( 1 + dp[N-1], 1 + dp[N-6], 1 + dp[N-9], 1 + dp[N-36], 1 + dp[N-81], ... )
    // dp[N]のひとつ前の状態というのは、
    // 1円、6円、9円、6^2円、9^2円、、、(N円より小さい6か9のべき乗円)しかない(引き出せる金額の選択肢がそれしかないので)
    // それらの遷移元から遷移してきたときのいくつかのパターンのなかで最小になったものがdp[N]のときの最小回数となる


    var dp = Array(110000) {N}
    dp[0] = 0

    for (i in 1..N) {
        var pow6 = 1
        while (pow6 <= i && pow6 <= N) {
            dp[i] = Math.min(dp[i], 1 + dp[i - pow6])
            pow6 *= 6
        }

        var pow9 = 1
        while (pow9 <= i && pow9 <= N) {
            dp[i] = Math.min(dp[i], 1 + dp[i - pow9])
            pow9 *= 9
        }
    }

    println(dp[N])
}